ocean/helium/snapshots.py

```diff
@@ -131,10 +136,267 @@ def download_file_from_snapshot(src, dst, snapshothash):
+    pass
+
+
+def dereference_physical_key(physical_key):
+    ty = physical_key['type']
+    if ty == 'local_file':
+        return open(physical_key['path'])
+
+    raise NotImplementedError
+
+
+class Package(object):
+    """ In-memory representation of a package """
+
+    def __init__(self, data=None, meta=None):
+        """
+        _data is of the form {logical_key: entry}
+        entry is of the form (physical_key, hash, size, user_meta)
+        physical_keys is a list of objects of the form {
+            schema_version: string
+            type: string
+            uri: string
+        }
+        hash is of the form {
+            type: string (e.g. "SHA256")
+            value: string
+        }
+        size is the length of the object in bytes
+        """
+        self._data = data or {}
+        self._meta = meta or {}
+
+    def _clone(self):
+        """
+        Returns clone of this package.
+        """
+        return Package(copy.deepcopy(self._data), copy.deepcopy(self._meta))
+
+    def __contains__(self, logical_key):
+        """
+        Checks whether the package contains a specified logical_key.
+
+        Returns:
+            True or False
+        """
+        return logical_key in self._data
+
+    @staticmethod
+    def load(path):
+        """
+        Loads a package from a path.
+
+        Args:
+            path: string representing the location to load the package from
+
+        Returns:
+            a new package object
+
+        Raises:
+            file not found
+            json decode error
+            invalid package exception
+        """
+        data = {}
+        with jsonlines.open(path) as reader:
+            meta = reader.read()
+            for obj in reader:
+                lk = obj['logical_key']
+                if lk in data:
+                    raise PackageException("Duplicate logical key while loading package")
+                del obj['logical_key']
+                data[lk] = obj
+
+        return Package(data, meta)
+
+    @staticmethod
+    def pkg(path):
+        """
+        Takes a package of a provided path.
+
+        Recursively enumerates every file in path, and returns a new
+        package that contains all those files.
+
+        Args:
+            path: path to package
+
+        Returns:
+            A new package of that path
+
+        Raises:
+            when path doesn't exist
+        """
+        # TODO: anything but local paths
+        # TODO: deserialization metadata
+        data = {}
+        meta = {}
+        src_path = pathlib.Path(path)
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+            entry = {
+                'hash': hash_file(f),
+                'size': os.path.getsize(f),
+                'user_meta': {},
+                'physical_keys': [{
+                    'type': 'local_file',
+                    'path': os.path.abspath(f)
```

Comment:
Should probably be relative to `path`, or some other root?

Reply:
Why wouldn't it be absolute? This is for the physical key